{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Yay Bioinformatics!\n",
    "\n",
    "Instructors:\n",
    "\n",
    "* Olga Botvinnik, PhD\n",
    "* Emily Wheeler\n",
    "* Alain Domissy\n",
    "\n",
    "## Bioinformatics homework\n",
    "\n",
    "1. How many people started the course?\n",
    "1. How many people finished?\n",
    "1. What did you think?\n",
    "\n",
    "### Python for Genomic Data Analysis\n",
    "### Command Line Tools for Bioinformatics\n",
    "\n",
    "## Stickie notes\n",
    "\n",
    "We will be using stickie notes (blue for done, red for still working) to help us instructors scan the room and see how the groups are doing.\n",
    "\n",
    "## Slack channel\n",
    "\n",
    "Throughout the course, the best way to reach us bioinformatics instructors will be via Slack (cshl-singlecell-2017.slack.com). It'll also be helpful for asking questions during lectures or hands-on stuff. I also highly encourage you to chat with your partner there!\n",
    "\n",
    "\n",
    "## Why single-cell RNA-Seq? Why not proteomics, genomics, etc?\n",
    "\n",
    "While we are primarily discussing single-cell RNA-Seq analyses, many of the same concepts can be applied to other single-cell 'omics methods: genomics, epigenomics, proteomics, metabolomics, etc. We're merely using RNA-seq as an example, and so you don't have to keep switching units throughout the course (minimizes cognitive overload).\n",
    "\n",
    "## Why Python 3? Why Jupyter Notebooks?\n",
    "\n",
    "We will be using Python as it is a very easy language to learn and has a rich library of scientific and machine learning resources far beyond biology, due to its ease of use. Thus there are many many machine learning libraries that we can take advantage of. It's not as bio-specific as the R/Bioconductor ecosystem, but in my opinion, you can always jump to R/Bioconductor when necessary, and read the output files in Python.\n",
    "\n",
    "We will be using Python 3 because you are the *future* of science, not the past. Python 3 has much better memory management (important for huge datasets of single cells!) and has been around for almost a decade, biut many scientists have not yet switched. Python 2 is very entrenched in the scientific world, where many labs have piles of code in Python 2 and won't update because it takes time to update your code.\n",
    "\n",
    "We will be using Jupyter Notebooks (rather than a studio environment like RStudio/MATLAB/Spyder) for the analyses because they will document your entire thought process throughout the exploratory and learning phase. This way you have files on your computer that you can take home and reference in the future. In Jupyter, you can take notes in plain text and remind yourself of why you were doing this in the first place :)\n",
    "\n",
    "## Your project: Reproduce a paper\n",
    "\n",
    "You and your partner will pick a paper of interest and as we learn methods of analysis, reproduce the figures from the paper as best you can. You will present your results at the final student data presentation.\n",
    "\n",
    "Homework until next time: Pick a paper and download its data (counts/expression matrix - not raw fastqs). Look for the \"Accession\" number in the paper and google it.\n",
    "\n",
    "## Schedule of topics\n",
    "\n",
    "1. (Emily) Where did this count matrix come from?\n",
    "    1. Overview of single-cell RNA-Seq analysis pipeline\n",
    "    1. Hands on with `cellranger` pipeline (Alain)\n",
    "1. **“Fishing expedition:” Take a tissue, dissociate it → ?? → subpopulations!**\n",
    "   1. **Paper:** Macosko - 2015 - “Highly Parallel Genome-wide Expression Profiling of Individual Cells Using Nanoliter Droplets” (aka “the dropseq paper”)\n",
    "https://www.ncbi.nlm.nih.gov/pubmed/26000488\n",
    "   2. **Optional Reviews:**\n",
    "Bacher and Kendziorski - 2016 - “Design and computational analysis of single-cell RNA-sequencing experiments”\n",
    "https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0927-y\n",
    "Kolodziejczyk et al - 2015 - “The Technology and Biology of Single-Cell RNA Sequencing”\n",
    "http://www.cell.com/molecular-cell/fulltext/S1097-2765(15)00261-0\n",
    "    3. *Question: How does gene dropout affect your results? How can you use computational methods to deal with dropout?*\n",
    "        1. Relationships between cells\n",
    "            1. Linear vs nonlinear correlation\n",
    "            1. Distance metrics\n",
    "            1. Hierarchical clustering\n",
    "        1. Interactive exploration of dropout and clusteirng\n",
    "        4. Reproduce Figure 5\n",
    "            1. A hint of compressed sensing/Robust PCA\n",
    "1. **Case vs Control**\n",
    "   1. **Paper:** Segerstolpe - 2016 - “Single-Cell Transcriptome Profiling of Human Pancreatic Islets in Health and Type 2 Diabetes” http://www.cell.com/cell-metabolism/fulltext/S1550-4131(16)30436-3\n",
    "   2. **Optional Review:**\n",
    "        Stegle et al - 2015 - “Computational and analytical challenges in single-cell transcriptomics”\n",
    "        https://www.ncbi.nlm.nih.gov/pubmed/25628217\n",
    "    3. *Question: How can you find structure in high dimensional data?*\n",
    "        1. Finding structure in your data: Clustering\n",
    "            1. Smushing your data via linear vs nonlinear dimensionality reduction (PCA vs TSNE)\n",
    "            1. Finding groups of cells via linear vs nonlinear clustering (KMeans vs DBSCAN)\n",
    "            1. Interactive exploration of dimensionality reduction + clustering\n",
    "        1. Reproduce Figure 1\n",
    "3. **Molecular transformations over time (“pseudo-time”)** (Advanced)\n",
    "   1. **Paper:** Lönnberg et al -  2017  - “Single-cell RNA-seq and computational analysis using temporal mixture modelling resolves Th1/Tfh fate bifurcation in malaria”\n",
    "https://www.ncbi.nlm.nih.gov/pubmed/28345074\n",
    "   2. **Optional Reviews:**\n",
    "Canoodt et al - 2016 - “Computational methods for trajectory inference from single-cell transcriptomics”\n",
    "http://onlinelibrary.wiley.com/doi/10.1002/eji.201646347/epdf\n",
    "Symmons and Raj - 2016 - “What’s Luck Got to Do with It: Single Cells, Multiple Fates, and Biological Nondeterminism”\n",
    "https://www.ncbi.nlm.nih.gov/pubmed/27259209\n",
    "    3. *Question: How can cells be ordered along differentiation, e.g. from young to old, or early to late?*\n",
    "        1. Finding structure in your data: Regression\n",
    "            1. Regression demo\n",
    "            1. Discussion of \"psudeotime\" methods\n",
    "        1. Reproduce Figure 5\n",
    "4. **Perturbation** (Advanced)\n",
    "    1. **Paper:**\n",
    "    Adamson and Norman et al - 2016 - “A Multiplexed Single-Cell CRISPR Screening Platform Enables Systematic Dissection of the Unfolded Protein Response”\n",
    "    https://www.ncbi.nlm.nih.gov/labs/articles/27984733/\n",
    "    2. **Optional Review:**\n",
    "    Tanay and Regev - 2017 - “Scaling single-cell genomics from phenomenology to mechanism” https://www.ncbi.nlm.nih.gov/pubmed/28102262\n",
    "    Wagner and Klein - 2017 - “Genetic screening enters the single-cell era”\n",
    "    http://www.nature.com/doifinder/10.1038/nmeth.4196\n",
    "    3. *Question: *\n",
    "        1. Which genes actually matter?\n",
    "            1. Finding important genes via linear (SVM) and nonlinear (Decision Trees) classifiers\n",
    "            1. Compressed sensing: Robust PCA\n",
    "            1. Gene ontology\n",
    "        1. Reproduce Figure 3\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (cshl-sca-2017)",
   "language": "python",
   "name": "cshl-sca-2017"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
